### ISSUE TITLE
Do text operations: lowercase

### ISSUE THREAD

**Issue by feliksik** [dhall-lang/dhall-lang#631]
I would like to have a `let lowercase : Text -> Text`. 

Based on https://github.com/dhall-lang/dhall-lang/issues/164 I understand that Dhall aims to promote static typing instead of string comparisons , and hence does not offer string equality. So far so good. 

Now https://github.com/dhall-lang/dhall-lang/issues/164#issuecomment-392721414 mentions I can implement this externally. That would be perfect. 

I can load environment variables as dhall method, but how can I define a function implementation? It seems that I need to call some external program or something.

**Comment by Gabriel439** [comment]
@feliksik: See: https://github.com/dhall-lang/dhall-lang/wiki/How-to-add-a-new-built-in-function

... which in turn currently links to:

https://hackage.haskell.org/package/dhall/docs/Dhall-Tutorial.html#g:21

**Comment by feliksik** [comment]
Thank you @Gabriel439 . It seems that the instruction you mention allows me to build a custom dhall executable with extra functions; is that correct? 

https://github.com/dhall-lang/dhall-lang/issues/164#issuecomment-392721414 says: 

> you can simply supply ∀(a : Text) → ∀(b : Text) → Bool as an argument in your program

I thought this means one can use the vanilla Dhall executable, and load a 'plugin' as an argument or ENV variable. Is this possible? Or is the author really talking about the internal Haskell API `Dhall.inputWithSettings` that can be given an extra argument? (I could eventually manage this, but it would unfortunately make it a lot less accessible for me and harder to sell Dhall to my colleagues).

**Comment by singpolyma** [comment]
The proposal there is to have your Dhall expression take your needed function as an argument, implement the function in the host language, and pass the function as an argument when calling the Dhall function.

This will work if you're using a native Dhall implementation in your program, but won't help you if using a fallback tool such as dhall-to-yaml

**Comment by Gabriel439** [comment]
@feliksik: Yeah, the fundamental issue is that adding a new-builtin to the Haskell implementation requires writing new Haskell code and Haskell is a compiled language

However, I'd like to note that technically adding a `lowercase` built-in doesn't violate the precept of forbidding runtime string equality checks.  So depending on how people feel we could just standardize support for `Text/lowercase`/`Text/uppercase` built-ins.

**Comment by singpolyma** [comment]
I'm curious what the use case is, and if it could be solved with a better-typed design

**Comment by philandstuff** [comment]
I agree with @singpolyma: I'd prefer to keep the language small by default, and only add new features where there is a strong motivating use case.

I'd also add that `Text/lowercase` and `Text/uppercase` have a certain can-of-worms quality once you start considering Unicode and locales: for example, should `Text/uppercase "i"` be `"I"` or `"İ"`? Maybe in Turkey it should be the latter?  We should, where possible, delegate to a Unicode upper/lowercasing algorithm, but we need to choose if it should be locale-aware.  If yes, then the same Dhall code will have different behaviour on different systems, which seems bad.  If no, then we aren't really implementing Unicode properly, which seems bad.  One could square this circle by allowing users to specify locale as a parameter to `Text/uppercase` but then the language is starting to get complicated. Or we could stick with pretending that our Unicode string is actually 7-bit ASCII and ignoring higher code points.

**Comment by ari-becker** [comment]
I agree with @philandstuff here. There are additional issues where the Unicode text includes multiple languages, each with its own notion of what "uppercase" or "lowercase" might be (if at all - should calling `Text/lowercase` on a language with no notion of case, like Hebrew, return the same `Text` or should it return a type error saying that the `Text` in question is un-lowercase-able?).

We've run into this issue before (taking user input in the form of a dhall expression which may or may not be completely lowercase, when downstream configuration needs it to be lowercase) and we deal with it by scripting the issue away. Sure, it would be more convenient to have a `Text/lowercase` but I just don't see a satisfactory way of specifying it.

**Comment by Profpatsch** [comment]
Oh heck, please don’t add any Unicode processing to dhall

**Comment by feliksik** [comment]
> I'm curious what the use case is, and if it could be solved with a better-typed design

My use case is the following: I made a model that defines a prometheus alerting rules *and* some strongly typed fields that describe how to act on the alert ( `{ normalBehavior: Text, problem: Text, impact: Text, actionableSuggestions: Text, more: Text }` ). From this, I then generate 1) a Markdown runbook that is hosted on github pages, containing all the alertNames as section titles, which can be referenced with an anchor link. 2) the prometheus rules yml file, with an annotation that links to the hosted runbook with details on the alert. 

But github pages lowercases all html anchors associated with the title, hence my desire to lowercase the fields before templating the runbook sections. Obviously this is the consequence of the github idiosyncracy, and my workaround is just to accept that I should use lowercase alert names. 

I think it *would* be interesting to be able to do some text processing in Dhall, but I think adding things to the stdlib may lead to a never-ending story: tomorrow someone will want to replace slashes with `_`, and then `-` with `_`, because some tools don't allow such characters in certain places. 

I suspect I will run into such issues more often in the future, but this is speculative. I am very much in the process of figuring out best practices for Dhall, and where *not* to use it.

**Comment by feliksik** [comment]
To have a productive dialogue, I do think it's very important to carefully separate the following themes: 

**a)** does it make sense to *implement particular text transformation algorithms in Dhall*? (Many people here say "no, unicode, horror" -- I do not argue with this). 

**b)** does it make sense to somehow allow to build text transformations at all (in a plugin/module fashion, or in some other way) *outside of the core language*? (The user can then choose the unicode assumptions he likes). 

**c)** if b is worth considering, is it then desirable to (artificially) constrain such transformations to type `Text -> Text`, in particular to prevent `equals : Text -> Text -> Bool`. It sounds like a pretty artificial constraint, but if you are concerned the untyped crap people will start creating may indeed pollute your entire language reputation, I can imagine you are conservative here. 

### custom, non-core functions 
For b), *if* the answer would be "yes" (and I don't know whether it should be) -- I am brainstorming with the following options: 

* User builds his own dhall binary, with the custom functions added. This seems to be the current option. 

* introduce a type `CodePoint`, and functions `split : Text -> List CodePoint`, `concat: List CodePoint -> Text`, and `== : CodePoint -> CodePoint -> Boolean`. This would allow defining custom character mapping with a `map : { from : CodePoint, to : CodePoint }`, even though this is not a way to support entire unicode. At least, it's now up to the user, and you can use the mapping that is practical for the character-constraints of you devops tool. This looks like conceptually the cleanest option, but it could stretch the power of Dhall further than you consider desirable -- e.g. point c) above, you can not implement Text equality. 

Then there are the following brainfarts you should only read when in an optimistic mood: 
* Dhall could support a syntax like `let lowerCase = ./lowerCaseModule.o as Module`, where the compiled Haskell modules is dynamically loaded. You could also share these over http. This feels fancy and powerful, but I think this may be pretty bad from a security/side-effects perspective (even when you can verify hashes, loading a module is much more risky now than loading a Dhall file). 

* have some IPC/RPC mechanism (local pipe or tcp socket) where you can write values to a stream and read results back. This is probably implementable, but sounds like you'd really need a strong motivation to support this custom-functions thing. The moment Dhall implements SOAP, poor Gabriel will definitely realize his opus magnum has been wasted... ;-)

**Comment by singpolyma** [comment]
IMHO the "plugin/extension" model is the best for the current phase of Dhall's life. This can be done by adding a builtin, but that can often be avoided because functions are first-class in Dhall and some (most? all?) implementations allow passing native functions of the host language as arguments to Dhall functions.

If one is only using dhall-to-yaml or similar, instead of a full Dhall binding, this is of course trickier. It *might* be interesting to allow dhall-to-yaml to load plugins of some nature to enable something similar, which would be specific to the users of that tool. Would depend on how many people who want advanced Dhall uses also cannot use a full Dhall binding for their use case.

**Comment by Gabriel439** [comment]
@singpolyma makes a good point that a native language binding could support marshaling higher-order functions (i.e. marshaling in a function whose input(s) are functions from the host language).  In fact, the Haskell implementation is actually not far away from being able to do that because the new faster normalizer has a place in the syntax tree where it can store an arbitrary Haskell function for this purpose:

https://github.com/dhall-lang/dhall-haskell/blob/e0bd21e25d0ab2445e67d71cf3968d8f0ec8cf62/dhall/src/Dhall/Eval.hs#L153

In other words, you could write a Dhall function like this:

```dhall
-- ./example.dhall
λ(Text/equal : Text → Text → Bool) → Text/equal "ABC" "DEF"
```

... and then you marshal that into, say, Haskell, as a Haskell function of type:

```haskell
higherOrderFunction <- input auto "./example.dhall" :: IO ((Text -> Text -> Bool) -> Bool)
```

... which you could then use like this:

```haskell
higherOrderFunction (==)  -- = False
```

**Comment by Profpatsch** [comment]
We definitely need support for importing dhall higher-order functions to Haskell

**Comment by jneira** [comment]
,,

**Comment by Gabriel439** [comment]
@Profpatsch: So I took a stab at this and it's actually harder than I thought.  The issue is that once you add a constructor like this to the `Expr` type:

```haskell
data Expr s a =
    …
    | Function (Expr s a -> Expr s a)
```

... several instances and functions for `Expr` no longer work.  For example, you can't resolve imports in a separate phase, so a lot of other code would have to be refactored.

**Comment by singpolyma** [comment]
Could you have `Expr` and `NormalizationExpr` or something to model that these functions are only valid when constructing an application to normalize?

**Comment by Gabriel439** [comment]
@singpolyma: We already do.  `NormalizationExpr` is `Dhall.Eval.Val`.  The hard part is that we don't expose `NormalizationExpr` in the public API yet, and even if we did we'd have the same problem.  Specifically, `NormalizationExpr` only interprets an expression that has already been resolved (which is what `Expr` handles).

**Comment by shinzui** [comment]
Support for marshaling higher-order functions will help a lot around the lack of Text utilities. I landed on this issue while trying to find a way to compare a string from an environment that I want to convert into a union. I wonder if you're open to either supporting marshaling higher-order functions or comparing text.

**Comment by ari-becker** [comment]
> while trying to find a way to compare a string from an environment that I want to convert into a union.

There are creative ways to achieve this, i.e. in bash:

```bash
myvar='Foo' # could be Foo or Bar
dhall <<< "let FromEnv = < Foo | Bar > in FromEnv.${myvar}" # will normalize to < Foo | Bar >.Foo
```

**Comment by Profpatsch** [comment]
In general, dhall doesn’t aim to support all use-cases, but aims to give the user power to define their own semantics layer.
There’s multiple ways a `lowercase` function could be implemented, depending on how the dhall code is called.

Personally, I tend to use `dhall-to-nix`, so I can use the nicest integration, which is just functions:

```
\(lowercase: Text -> Text) -> lowercase "FOO"
````

which can just be called like a normal function in the nix code, to be passed the lowercasing function on the nix side. This is the “marshalling of higher-order functions’ idea from above.

If that’s not possible, you can create a semantic layer around your text:

```
let ModText = < Lowercased : Text | AsIs : Text >
in {
  asis = ModText.AsIs "stay BIG",
  lower = ModText.Lowercased "GET SMALL"
}
```

and in then import that and convert all `AsIs` fields to lowercase as a post-processing step.

The last way is adding now built-in symbols to the dhall parsing code (extending the language on the fly), but I don’t recommend that, because then your dhall files are essentially bound to the code that parses them and normal dhall tools (e.g. `dhall-to-json`) cannot load them anymore.

**Comment by Gabriel439** [comment]
Personally I'm starting to lean towards towards proposing support for `Text` built-ins that do not permit introspection (See: https://discourse.dhall-lang.org/t/text-manipulation-functions/176/16?u=gabriel439) and `lowercase` is one such function that would be safe according to that criterion.

**Comment by ad-si** [comment]
Since the replace function is now included in Dhall, why can't we add a `Text.toAsciiLowerCase` like this?
This avoids any Unicode problems, but should be good enough for 90% of use cases.
I needed it to auto generate slugs, where the parts of the slug were already ASCII only.

```dhall
let toAsciiLowerCase
    : Text -> Text
    = \(text : Text) ->
        (Prelude.Text.replace "A" "a"
        (Prelude.Text.replace "B" "b"
        (Prelude.Text.replace "C" "c"
        (Prelude.Text.replace "D" "d"
        (Prelude.Text.replace "E" "e"
        (Prelude.Text.replace "F" "f"
        (Prelude.Text.replace "G" "g"
        (Prelude.Text.replace "H" "h"
        (Prelude.Text.replace "I" "i"
        (Prelude.Text.replace "J" "j"
        (Prelude.Text.replace "K" "k"
        (Prelude.Text.replace "L" "l"
        (Prelude.Text.replace "M" "m"
        (Prelude.Text.replace "N" "n"
        (Prelude.Text.replace "O" "o"
        (Prelude.Text.replace "P" "p"
        (Prelude.Text.replace "Q" "q"
        (Prelude.Text.replace "R" "r"
        (Prelude.Text.replace "S" "s"
        (Prelude.Text.replace "T" "t"
        (Prelude.Text.replace "U" "u"
        (Prelude.Text.replace "V" "v"
        (Prelude.Text.replace "W" "w"
        (Prelude.Text.replace "X" "x"
        (Prelude.Text.replace "Y" "y"
        (Prelude.Text.replace "Z" "z" text)
        )))))))))))))))))))))))))
```

**Comment by Gabriel439** [comment]
@ad-si: What about adding a primitive that could also lowercase Unicode, too?  I don't think it would be much more difficult to standardize than the ASCII-only version.

**Comment by ad-si** [comment]
@Gabriel439 Sure, if it's really not much more difficult!
It just sounded like a Unicode compatible version would be too much effort and that's why it has been postponed. An ASCII only version could have been a good stepping stone.

**Comment by Gabriel439** [comment]
The only reason I haven't standardized a bunch of stuff recently is because I've been slowly mechanizing the standard first, to make it easier for both new and existing implementations to follow the standard (See: https://github.com/dhall-lang/dhall-lang/issues/959).  Other than that, there's nothing really blocking this other than somebody taking the time to standardize it

**Comment by MonoidMusician** [comment]
Unicode casing is a little complicated, because there are both simple (codepoint-by-codepoint) and [full](https://www.unicode.org/Public/13.0.0/ucd/SpecialCasing.txt) algorithms (which may replace single codepoints by multiple codepoints), and there's four different case conversions available: uppercase, lowercase, titlecase, and case folding (which is often lowercase, but sometimes uppercase – it's meant for text comparisons).

For titlecase in particular, I wasn't able to find an actual reference to specify how to split strings into words in order to titlecase their components. It also depends on the language, more than the other algorithms do. (Uppercase/lowercase only have language exceptions for Turkish dotted/dotless “i”, I believe.) And there are a few weird [edge cases with normalization forms](https://djangocas.dev/blog/python-unicode-string-lowercase-casefold-caseless-match/#unicode-default-caseless-matching) (NF(K){C,D}), but that doesn't affect most applications, only caseless matching.

You can see my experience with it here: https://github.com/purescript-contrib/purescript-unicode/pull/15 (I had to add parsing for two more files from the unicode standard, and opted not to implement titlecasing for strings, just for codepoints)

So is it probably okay to just formalize the full uppercase and lowercase algorithms. That's certainly “good enough” for most applications. But there's definitely some subtleties here, depending on what you really want to accomplish …

**Comment by Gabriel439** [comment]
Oh, I guess it is trickier than I thought.  I just did a bit more research and apparently uppercasing and lowercasing a string properly in Unicode is locale-sensitive.  The one thing that appears to be simple is case-folding (for the purpose of case-insensitive string comparisons).

In that case, we could implement a Prelude utility which handles just the ASCII case like @ad-si suggested until we can figure out how to handle the general case

**Comment by Gabriel439** [comment]
I put up a pull request to implement the ASCII version in the Prelude: https://github.com/dhall-lang/dhall-lang/pull/1164

**Comment by jsoref** [comment]
For anyone reading this, `Prelude.Text.lowerASCII` and `Prelude.Text.upperASCII` have been available since  https://github.com/dhall-lang/dhall-lang/releases/tag/v20.2.0

**Comment by TristanCacqueray** [comment]
Thanks @jsoref !
